### DISCUSSION TITLE
Measuring MLE SendAnnounce time

### DISCUSSION THREAD

**Discussion by rretanubun** [openthread/openthread#8662] [Q&A] [answered]
# Confirming my understanding
- As part of `MLE`, the thread stack occasionally (`OPENTHREAD_CONFIG_ANNOUNCE_SENDER_INTERVAL` controlled) sends announcements using `MLE:SendAnnounce()`
- Depending on `channel mask` this would traverse and potentially send announcements to all channels.
- During this time, because the radio is momentarily used for another channel, some packets may be missed.
- If I want to measure the time that the radio is unavailable for the active channel, I can record timestamp measurements here:
https://github.com/openthread/openthread/blob/main/src/core/thread/mle.cpp#L2198-2200

For example:
```
    t1 = get_uptime_in_uSec();
    SuccessOrExit(error = message->SendTo(aDestination));
    t2 = get_uptime_in_uSec();

    LogInfo("Send Announce on channel %d [%llu]", aChannel, (t2 - t1));
```

# Question
1. Is my understanding correct? or is there a different spot in code to measure?
2. Other than modifying OPENTHREAD_CONFIG_ANNOUNCE_SENDER_INTERVAL  is there a network event that can be triggered to cause the Announcement to go out earlier (to aid in collecting statistical data for example).

Thank you for everyone's time.

**Comment by jwhui** [comment, marked as answer]
> Is my understanding correct? or is there a different spot in code to measure?

The `SendTo()` method simply schedules the packet for transmission and does not actually wait for the radio to complete transmission. If you want the most accurate time stamping, you'll probably want to do that in the radio driver itself. See the `otPlatRadioTransmit` and `otPlatRadioTxDone` platform APIs.

> Other than modifying OPENTHREAD_CONFIG_ANNOUNCE_SENDER_INTERVAL is there a network event that can be triggered to cause the Announcement to go out earlier (to aid in collecting statistical data for example).

There isn't a network event that triggers it. Note, however, that the announce messages may be suppressed if there are a number of other neighbors sending the same message. You can find the configurations in [src/core/config/announce_sender.h](https://github.com/openthread/openthread/blob/main/src/core/config/announce_sender.h).

**Reply by rretanubun** [reply]
Ah okay. I am trying to do this by taking timestamps and printing to console. if I go to the radio driver, there is a lot of noise since the radio driver is at a much lower level and handles all kinds of transmissions, no?

Is there a way to get (maybe not the most accurate, but practical) timestamp of the time the radio is sending announcements on the non-operational channel?

**Reply by jwhui** [reply]
You could print the message only when sending announce messages. You can identify announce messages when the Destination PAN ID is set to the broadcast pan (`0xffff`) or the MAC Security Mode is set to 2.
